src/loaders/anndata-loaders/MatrixZarrLoader.js

```diff
@@ -1,66 +1,236 @@
+    if (geneFilterZarr) {
+      geneFilter = await this.getFlatArrDecompressed(geneFilterZarr);
+    }
+    const { _index } = await this.getJson('var/.zattrs');
+    if (this.geneNames) {
+      return this.geneNames;
+    }
+    this.geneNames = this.getFlatTextArr(`var/${_index}`).then(data => data.filter((_, j) => !geneFilter || geneFilter[j]));
+    return this.geneNames;
+  }
+
+  /**
+   * Class method for loading a filtered subset of the genes list
+   * @param {String} filterZarr A location in the zarr store to fetch a boolean array from.
+   * @returns {Array} A list of filtered genes.
+   */
+  async _getFilteredGenes(filterZarr) {
+    const filter = await this.getFlatArrDecompressed(filterZarr);
+    const geneNames = await this.loadGeneNames();
+    const genes = geneNames.filter((_, i) => filter[i]);
+    return genes;
+  }
+
+  /**
+   * Class method for getting the integer indices of a selection of genes within a list.
+   * @param {Array} selection A list of gene names.
+   * @returns {Array} A list of integer indices.
+   */
+  async _getGeneIndices(selection) {
+    const geneNames = await this.loadGeneNames();
+    return selection.map(gene => geneNames.indexOf(gene));
+  }
+
+  /**
+   * Class method for getting the number of cells i.e entries in `obs`.
+   * @returns {Number} The number of cells.
+   */
+  async _getNumCells() {
+    const cells = await this.loadCellNames();
+    return cells.length;
+  }
+
+  /**
+   * Class method for getting the number of genes i.e entries in `var`,
+   * potentially filtered by `genesFilter`.
+   * @returns {Number} The number of genes.
+   */
+  async _getNumGenes() {
+    const cells = await this.loadGeneNames();
+    return cells.length;
+  }
+
+  /**
+   * Class method for opening the sparse matrix arrays in zarr.
+   * @returns {Array} A list of promises pointing to the indptr, indices, and data of the matrix.
+   */
+  async _openSparseArrays() {
+    const {
+      options: { matrix },
+      store,
+    } = this;
+    if (this.sparseArrays) {
+      return this.sparseArrays;
+    }
+    this.sparseArrays = Promise.all(
+      ['indptr', 'indices', 'data'].map(name => openArray({ store, path: `${matrix}/${name}`, mode: 'r' })),
+    );
+    return this.sparseArrays;
+  }
+
+  /**
+   * Class method for loading a gene selection from a CSC matrix.
+   * @param {Array} selection A list of gene names whose data should be fetched.
+   * @returns {Promise} A Promise.all array of promises containing Uint8Arrays, one per selection.
+   */
+  async _loadCSCGeneSelection(selection) {
+    const indices = await this._getGeneIndices(selection);
+    const [indptrArr, indexArr, cellXGeneArr] = await this._openSparseArrays();
+    const numCells = await this._getNumCells();
+    const { data: cols } = await indptrArr.getRaw(null);
+    // If there is not change in the column indexer, then the data is all zeros
+    return Promise.all(
+      indices.map(async (index) => {
```

Comment:
Would you be able to extract this logic to a separate function and add tests for CSC and CSR?. For instance, given a toy sparse array with `indptrArr, indexArr, cellXGeneArr`, can we test that the data loaded for a particular `selection` corresponds to the expected gene? I think it may be important to just do a sanity check to verify that (after hypothetical future code maintenance) we don't ever show the user gene expression data which corresponds to the wrong gene or cell